nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/ConsumePulsar_1_X.java

```diff
@@ -0,0 +1,206 @@
+
+@Tags({"Pulsar", "Get", "Ingest", "Ingress", "Topic", "PubSub", "Consume"})
+@CapabilityDescription("Consumes messages from Apache Pulsar "
+        + "The complementary NiFi processor for sending messages is PublishPulsar.")
+@InputRequirement(InputRequirement.Requirement.INPUT_FORBIDDEN)
+public class ConsumePulsar_1_X extends AbstractPulsarConsumerProcessor {
+
+    private static final List<PropertyDescriptor> PROPERTIES;
+    private static final Set<Relationship> RELATIONSHIPS;
+
+    static {
+        final List<PropertyDescriptor> properties = new ArrayList<>();
+        properties.add(PULSAR_CLIENT_SERVICE);
+        properties.add(TOPIC);
+        properties.add(SUBSCRIPTION);
+        properties.add(ASYNC_ENABLED);
+        properties.add(MAX_ASYNC_REQUESTS);
+        properties.add(ACK_TIMEOUT);
+        properties.add(PRIORITY_LEVEL);
+        properties.add(RECEIVER_QUEUE_SIZE);
+        properties.add(SUBSCRIPTION_TYPE);
+        properties.add(MAX_WAIT_TIME);
+
+        PROPERTIES = Collections.unmodifiableList(properties);
+
+        final Set<Relationship> relationships = new HashSet<>();
+        relationships.add(REL_SUCCESS);
+        RELATIONSHIPS = Collections.unmodifiableSet(relationships);
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return RELATIONSHIPS;
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return PROPERTIES;
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+
+        try {
+            if (context.getProperty(ASYNC_ENABLED).asBoolean()) {
+                // Launch consumers
+                consumeAsync(context, session);
+
+                // Handle completed consumers
+                handleAsync(context, session);
+
+            } else {
+                consume(context, session);
+            }
+        } catch (PulsarClientException e) {
+            getLogger().error("Unable to consume from Pulsar Topic ", e);
+            context.yield();
+            throw new ProcessException(e);
+        }
+
+    }
+
+    private void handleAsync(ProcessContext context, ProcessSession session) {
+
+        try {
+            Future<Message> done = consumerService.take();
+            Message msg = done.get();
+
+            if (msg != null) {
+                FlowFile flowFile = null;
+                final byte[] value = msg.getData();
+                if (value != null && value.length > 0) {
+                    flowFile = session.create();
+                    flowFile = session.write(flowFile, out -> {
+                        out.write(value);
+                    });
+
+                   session.getProvenanceReporter().receive(flowFile, "From " + getWrappedConsumer(context).getTransitURL());
+                   session.transfer(flowFile, REL_SUCCESS);
+                   session.commit();
```

Comment:
You might want to consider doing this every X number of flowfiles or based on a timer.